perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/BBMRIResources.java

```diff
@@ -0,0 +1,292 @@
+import cz.metacentrum.perun.registrar.model.ApplicationFormItemData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Registration module for BBMRI Collections, Networks, canSERV resources, etc.
+ * Module:
+ * 1. reads input with IDs of the resources and checks, whether groups representing resources exist
+ *   - uses the content of attribute RESOURCE_ID_ATTR_NAME as the name of the attribute containing IDs of the
+ *     resources on groups
+ *   - parses the root group, under which the groups representing specified resources are located
+ *     - if RESOURCE_ORIGIN_ENABLED_ATTR_NAME is set to true, the root group will be parsed from input from the form
+ *       (where input is select named "resourceOrigin")
+ *     - else if RESOURCES_ROOT_GROUP_ATTR_NAME is set, the root group will be parsed from the value of this attribute
+ *     - else the root group will be considered the target group of the application
+ * 2. adds users to the appropriate groups
+ *
+ * @author Dominik Frantisek Bucik <bucik@ics.muni.cz> (modifications)
+ */
+public class BBMRIResources extends DefaultRegistrarModule {
+
+	private final static Logger log = LoggerFactory.getLogger(BBMRIResources.class);
+
+	// field names
+	private static final String RESOURCE_IDS = "resourceIds";
+	private static final String RESOURCE_ORIGIN = "resourceOrigin";
+
+	// configuration attributes
+	private static final String RESOURCE_ID_ATTR_NAME = "urn:perun:group:attribute-def:def:resourceIDAttrName";
+	private static final String RESOURCE_ORIGIN_ENABLED_ATTR_NAME = "urn:perun:group:attribute-def:def:resourceOriginEnabled";
+	private static final String RESOURCES_ROOT_GROUP_ATTR_NAME = "urn:perun:group:attribute-def:def:resourcesRootGroup";
+
+	/**
+	 * Find groups representing resources by input. Groups are looked either based on the input,
+	 * or by configuration via attribute, or for in subgroups of group the module is assigned to.
+	 * Add user as a member to the groups and remove from this group.
+	 *
+	 * @param session who approves the application
+	 * @param app application
+	 * @return unchanged application
+	 * @throws PerunException in case of internal error in Perun
+	 */
+	@Override
+	public Application approveApplication(PerunSession session, Application app)
+		throws VoNotExistsException, UserNotExistsException, PrivilegeException,
+		MemberNotExistsException, RegistrarException, GroupNotExistsException,
+		AttributeNotExistsException, WrongAttributeAssignmentException, ExternallyManagedException,
+		WrongAttributeValueException, WrongReferenceAttributeValueException, NotGroupMemberException
+	{
+		// get perun and beans from session
+		PerunBl perun = (PerunBl)session.getPerun();
+		Vo vo = app.getVo();
+		User user = app.getUser();
+		Member member = perun.getMembersManagerBl().getMemberByUser(session, vo, user);
```

Comment:
If you expect this to work only for the group applications when user is already a member of the VO, then you should probably check application type (if group is present) and also handle MemberNotExistsException in a better way.
Eg. you can silently skip it and do nothing or you can log the warning messages about it or throw better (custom) exception, which will be clear to the support team when they will be handling the error report.

Reply:
Think I have done my best how to really handle it. Thanks for the tip!